lite2/client.go

```diff
@@ -0,0 +1,304 @@
+// future.
+type TrustOptions struct {
+	// Only trust commits up to this old.
+	// Should be equal to the unbonding period minus a configurable evidence
+	// submission synchrony bound.
+	Period time.Duration
+
+	// Height and Hash must both be provided to force the trusting of a
+	// particular height and hash.
+	Height int64
+	Hash   []byte
+}
+
+type mode int
+
+const (
+	sequential mode = iota
+	bisecting
+)
+
+// Option sets a parameter for the light client.
+type Option func(*Client)
+
+// SequentialVerification option can be used to instruct Verifier to
+// sequentially check the headers. Note this is much slower than
+// BisectingVerification, albeit more secure.
+func SequentialVerification() Option {
+	return func(c *Client) {
+		c.mode = sequential
+	}
+}
+
+// BisectingVerification option can be used to instruct Verifier to check the
+// headers using bisection algorithm described in XXX.
+//
+// trustLevel - maximum change between two not consequitive headers in terms of
+// validators & their respective voting power, required to trust a new header
+// (default: 1/3).
+func BisectingVerification(trustLevel float32) Option {
+	if trustLevel > 1 || trustLevel < 1/3 {
+		panic(fmt.Sprintf("trustLevel must be within [1/3, 1], given %v", trustLevel))
+	}
+	return func(c *Client) {
+		c.mode = bisecting
+		c.trustLevel = trustLevel
+	}
+}
+
+// AlternativeSources option can be used to supply alternative providers, which
+// will be used for cross-checking the primary provider.
+func AlternativeSources(providers []provider.Provider) Option {
+	return func(c *Client) {
+		c.alternatives = providers
+	}
+}
+
+type Client struct {
+	chainID        string
+	trustingPeriod time.Duration
+	mode           mode
+	trustLevel     float32
+
+	// Primary provider of new headers.
+	primary provider.Provider
+
+	// Alternative providers for checking the primary for misbehavior by
+	// comparing data.
+	alternatives []provider.Provider
+
+	// Where trusted headers are stored.
+	trustedStore  store.Store
+	trustedHeader *types.SignedHeader
+	trustedVals   *types.ValidatorSet
+
+	logger log.Logger
+}
+
+// NewClient returns a new light client.
+func NewClient(
+	chainID string,
+	trustOptions TrustOptions,
+	primary provider.Provider,
+	trustedStore store.Store,
+	options ...Option) (*Client, error) {
+
+	c := &Client{
+		chainID:        chainID,
+		trustingPeriod: trustOptions.Period,
+		mode:           bisecting,
+		trustLevel:     DefaultTrustLevel,
+		primary:        primary,
+		trustedStore:   trustedStore,
+		logger:         log.NewNopLogger(),
+	}
+
+	for _, o := range options {
+		o(c)
+	}
+
+	err := c.initializeWithTrustOptions(trustOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// TODO: cross check primary with alternatives providers
+// func (c *Client) crossCheckPrimary() {
+// }
+
+func (c *Client) initializeWithTrustOptions(options TrustOptions) error {
+	h, err := c.primary.SignedHeader(options.Height)
```

Comment:
TODO: check if we're ahead of options.Height. If so, revert?

Reply:
> Is initializeWithTrustOptions intended to be called multiple times?

no, one time. but light client can be started/stopped multiple times (due to crash or other reasons)